centrallix-sysdoc/OSDriver_Authoring.md

```diff
@@ -115,497 +170,1032 @@ License:  Copyright (C) 2001-2011 LightSys Technology Services. See LICENSE.txt
+
+- 📖 **Note**: This memory manager is usually the wrong choice for blocks of memory of arbitrary sizes.  It is intended for allocating structures quickly that are of a specific size.  For example, allocated space for a struct that is always the same size.
 
-In addition the memory manager provides statistics on the hit ratio of allocated blocks coming from the lists vs. malloc(), and information on how many blocks of each size/type are allocated out and cached.  This information can be invaluable in tracking down memory leaks.
+- 🥱 **tl;dr**: Use `nmMalloc()` for structs, not for strings.
 
-One caveat is that this memory manager does not provide a realloc() function, so the standard malloc(), free(), and realloc() must be used for blocks of memory that might grow in size.  This memory manager is also perhaps not the best to use for blocks of memory of arbitrary sizes, but rather is best for allocating structures quickly that are of a specific size and belong to specific objects, such as the StructInf structure or the SnNode structure, and others.  In short, use it for structures, but not for strings.
+- ⚠️ **Warning**: Calling `free()` on a block obtained from `nmMalloc()` or calling `nmFree()` on a block obtained from `malloc()` might not crash the program immediately.  Instead, it will result in either inefficient use of the memory manager, or a significant memory leak, respectively.  These practices will also lead to incorrect results from the statistics and block count mechanisms.
 
-Empirical testing has shown an increase of performance of around 50% or more in programs with the newmalloc module in use.
 
 The following are the functions for the newmalloc module:
 
-### void* nmMalloc(int size)
-This function allocates a block of the given 'size'.  It returns NULL if the memory could not be allocated.
+### nmMalloc()
+```c
+void* nmMalloc(int size);
+```
+This function allocates a block of the given `size`.  It returns `NULL` if the memory could not be allocated.
+
+
+### nmFree()
+```c
+void nmFree(void* ptr, int size);
+```
+This function frees the block of memory.
+
+- ⚠️ **Warning**: The caller **must know the size of the block.**  Getting this wrong is very bad!!  For structures, this is trivial, simply use `sizeof()`, exactly the same as with `nmMalloc()`.
+
+
+### nmStats()
+```c
+void nmStats(void);
+```
+Prints statistics about the memory manager, for debugging and optimizing.
+
+For example:
+```
+NewMalloc subsystem statistics:
+   nmMalloc: 0 calls, 0 hits (-nan%)
+   nmFree: 0 calls
+   bigblks: 0 too big, 0 largest size
+```
+
+<!-- TODO: Greg - In my personal testing, it always printed this same output.  Possibly some flags need to be enabled to get this to work properly? -->
```

Comment:
centrallix-lib needs to be built with the configure option --enable-debugging

Reply:
Noted, this section has been updated.